### DISCUSSION TITLE
Missing documentation for React.lazy and React.Suspense

### DISCUSSION THREAD

**Discussion by ackvf** [vercel/next.js#17979] [Help] [answered]
Here https://nextjs.org/docs/advanced-features/dynamic-import I am missing some mention of and differences between Next's own `dynamic()` and `React.lazy()`; and the loading component
`dynamic(..., { loading: CustomLoading })` and `<React.Suspense fallback={ CustomLoading } ... />`

From what I read in the very sparse online sources and next issues (such as https://github.com/vercel/next.js/issues/16516) , I can assume some basic differences between `dynamic` and `lazy`, but I still don't understand the difference between

```
const DynamicComponentWithCustomLoading = dynamic(
  () => import('../components/hello'),
  { loading: () => <p>...</p> }
)

const Home = () => (
  <DynamicComponentWithCustomLoading />
)
```
and
```
const DynamicComponent = dynamic(
  () => import('../components/hello')
)

const Home = () => (
  <React.Suspense fallback={<p>...</p>}>
    <DynamicComponent/>
  </React.Suspense>
)
```

One thing that comes to my mind immediately is that with Suspense, I can pass some props to the Loading component, but with `dynamic`, the component is defined outside of the rendering function's scope.

---

Should this be in a github issue instead? I would like to see the information in those docs permanently as I believe it is valuable information.

**Comment by jamesmosier** [comment, marked as answer]
The difference between `next/dynamic` and `React.lazy` is that `next/dynamic` supports code splitting for server side rendered applications while `React.lazy` does not. `next/dynamic` is very similar to [react-loadable](https://github.com/jamiebuilds/react-loadable) with some modification for Next's purposes. `React.Suspense` is also not supported for server side rendering.

From what I have seen with other sections of the docs, the Next docs do not go into detail about the difference between Next's API and other APIs - instead they just explain what the package/API does for the user of it. But as always if you think there is something missing about how `next/dynamic` works feel free to open a PR 🎉

**Reply by ackvf** [reply]
Yes, I feel there is something missing, but as I don't understand the possibilities of each and differences between them, creating a PR didn't feel like a good idea.

So, is `React.Suspense` and `React.lazy` discouraged for SSR apps, while it doesn't make any difference on client-side?
Are my two code snippets equivalent on client-side?

Given my one difference regarding the props, is there a way to pass the props to the Loading component while doing it the `next/dynamic` way? - Or should I use `React.Suspense` for it (while not breaking any Next.js compatibility)?

Is there actually an official subset of React features compatible with Next.js? I wasn't aware I should not use everything React offers to not break my Next.js app since Next.js is officially a React framework.

_PS: I unmarked this topic as answered as there are still some details I don't understand, so please, don't take it personally._

**Reply by leerob** [reply]
@jamesmosier has a great answer here. In general, you probably don't need Suspense. You also don't need `React.lazy` as he explains, because that's what `next/dynamic` is for.

When Suspense is generally recommended for wide usage, our documentation will be updated to reflect that. Right now, it would do more harm to mention it than help in my opinion.

**Reply by ackvf** [reply]
Thank you, now I can see it.

**Reply by wushaobo** [reply]
72d ago might become a reason I dare to raise a similar question again: what's your recommendation between `next/dynamic` and `@loadable/component`, the latter seems with the same purpose as the front, and [recommended by React](https://reactjs.org/docs/code-splitting.html#reactlazy). 

I am actually investigating on the proper strategies for handling the ChunkLoadError but not sure which of the two should be abandoned because I was using both...

------updates------
Well, I had got the suggestion from a member in an github issue. I turned to `next/dynamic` which is the `next.js` way of loadable. Consequently, ChunkLoadError has never happened so far in last 10 days across several times of deployment.

**Comment by abengit** [comment]
Some of the differences between next/dynamic and React.lazy 
- React.lazy does not allow SSR, while it is possible in next/dynamic
- React.lazy does not allow named exports, next/dynamic allow both named and default exports
- React.lazy typeof is a react.lazy while next/dynamic is defined in react as react.forward_ref

Also i would like to mention that you can import react components with next/dynamic or react.lazy , and both imports accept props as any React component
